File: packages/ctrl-kernel-sdk/src/effect.ts
// Effect — first-class side effect. Returned from actor handlers.
// L1 Kernel executes effects asynchronously, checking capability.

import type { Capability } from './capability.js';
import type { Event } from './event.js';

export type Effect =
  | LlmCallEffect
  | McpInvokeEffect
  | EmitEventEffect
  | SpawnActorEffect
  | PersistEventEffect
  | ShellExecEffect
  | HttpRequestEffect;

export interface LlmCallEffect {
  readonly kind: 'LlmCall';
  readonly model: string;
  readonly prompt: unknown;
  readonly deadlineMs: number;
  readonly replyTo: string; // ActorId
}

export interface McpInvokeEffect {
  readonly kind: 'McpInvoke';
  readonly server: string;
  readonly tool: string;
  readonly args: unknown;
  readonly replyTo: string;
}

export interface EmitEventEffect {
  readonly kind: 'EmitEvent';
  readonly target: string;
  readonly event: Event;
}

export interface SpawnActorEffect {
  readonly kind: 'SpawnActor';
  readonly prototype: string;
  readonly capability: Capability;
  readonly parentId: string;
  readonly initialState: unknown;
}

export interface PersistEventEffect {
  readonly kind: 'PersistEvent';
  readonly event: Event;
  readonly index?: ReadonlyArray<string>;
}

export interface ShellExecEffect {
  readonly kind: 'ShellExec';
  readonly cmd: string;
  readonly args: ReadonlyArray<string>;
  readonly replyTo: string;
}

export interface HttpRequestEffect {
  readonly kind: 'HttpRequest';
  readonly method: 'GET' | 'POST' | 'PUT' | 'DELETE';
  readonly url: string;
  readonly headers?: Record<string, string>;
  readonly body?: unknown;
  readonly replyTo: string;
}

// Builder helpers — concise effect construction in actor handlers.
export const Eff = {
  llmCall: (model: string, prompt: unknown, deadlineMs: number, replyTo: string): LlmCallEffect => ({
    kind: 'LlmCall', model, prompt, deadlineMs, replyTo,
  }),
  mcpInvoke: (server: string, tool: string, args: unknown, replyTo: string): McpInvokeEffect => ({
    kind: 'McpInvoke', server, tool, args, replyTo,
  }),
  emit: (target: string, event: Event): EmitEventEffect => ({
    kind: 'EmitEvent', target, event,
  }),
  spawn: (prototype: string, capability: Capability, parentId: string, initialState: unknown): SpawnActorEffect => ({
    kind: 'SpawnActor', prototype, capability, parentId, initialState,
  }),
  persist: (event: Event, index?: ReadonlyArray<string>): PersistEventEffect => ({
    kind: 'PersistEvent', event, index,
  }),
} as const;

File: src-tauri/src/kernel/effect.rs
// Effect — first-class side effect. Returned from actor handlers.
//
// Kernel executes effects asynchronously, checking capability before each.
// Effects carry `reply_to` ActorId for async result delivery.
// Mirrors @ctrl/kernel-sdk effect.ts.

use crate::kernel::actor::ActorId;
use crate::kernel::capability::Capability;
use crate::kernel::event::Event;
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "PascalCase")]
pub enum Effect {
    LlmCall {
        model: String,
        prompt: serde_json::Value,
        deadline_ms: u64,
        reply_to: ActorId,
    },
    McpInvoke {
        server: String,
        tool: String,
        args: serde_json::Value,
        reply_to: ActorId,
    },
    EmitEvent {
        target: ActorId,
        event: Event,
    },
    SpawnActor {
        prototype: String,
        capability: Capability,
        parent_id: ActorId,
        initial_state: serde_json::Value,
    },
    PersistEvent {
        event: Event,
        #[serde(default, skip_serializing_if = "Vec::is_empty")]
        index: Vec<String>,
    },
    ShellExec {
        cmd: String,
        args: Vec<String>,
        reply_to: ActorId,
    },
    HttpRequest {
        method: HttpMethod,
        url: String,
        #[serde(default)]
        headers: BTreeMap<String, String>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        body: Option<serde_json::Value>,
        reply_to: ActorId,
    },
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "UPPERCASE")]
pub enum HttpMethod {
    Get,
    Post,
    Put,
    Delete,
}

/// Executes effects asynchronously, checking capability + deadlines.
/// P2.1 skeleton — concrete execution wired in P2.4-P2.7.
pub struct EffectExecutor;

impl EffectExecutor {
    pub fn new() -> Self {
        Self
    }
}

impl Default for EffectExecutor {
    fn default() -> Self {
        Self::new()
    }
}
